token_test.go

```diff
@@ -0,0 +1,77 @@
+package jwt_test
+
+import (
+	"testing"
+
+	"github.com/golang-jwt/jwt"
+)
+
+func TestToken_SigningString(t1 *testing.T) {
+	type fields struct {
+		Raw       string
+		Method    jwt.SigningMethod
+		Header    map[string]interface{}
+		Claims    jwt.Claims
+		Signature string
+		Valid     bool
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		want    string
+		wantErr bool
+	}{
+		{
+			name: "",
+			fields: fields{
+				Raw:    "",
+				Method: jwt.SigningMethodHS256,
+				Header: map[string]interface{}{
+					"typ": "JWT",
+					"alg": jwt.SigningMethodHS256.Alg(),
+				},
+				Claims:    jwt.StandardClaims{},
+				Signature: "",
+				Valid:     false,
+			},
+			want: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30",
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t1.Run(tt.name, func(t1 *testing.T) {
+			t := &jwt.Token{
+				Raw:       tt.fields.Raw,
+				Method:    tt.fields.Method,
+				Header:    tt.fields.Header,
+				Claims:    tt.fields.Claims,
+				Signature: tt.fields.Signature,
+				Valid:     tt.fields.Valid,
+			}
+			got, err := t.SigningString()
+			if (err != nil) != tt.wantErr {
+				t1.Errorf("SigningString() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != tt.want {
+				t1.Errorf("SigningString() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func BenchmarkToken_SigningString(b *testing.B) {
```

Comment:
Did you do a comparison between the old and the new approach? It may be the case, that this new approach slightly increases number of allocations, since you add addtional two variables for `claims` and `header` whereas the original implementation directly stores it into an array (which you need to create temporarily as well in the last call)

Reply:
I took the liberty to test both implementations and there doesn't seem to be a perf regression:

current

> goos: darwin
> goarch: amd64
> pkg: github.com/golang-jwt/jwt/v4
> cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
> BenchmarkToken_SigningString
> BenchmarkToken_SigningString/BenchmarkToken_SigningString
> BenchmarkToken_SigningString/BenchmarkToken_SigningString-12         	  820801	      1417 ns/op	     600 B/op	      17 allocs/op
> PASS

new

> goos: darwin
> goarch: amd64
> pkg: github.com/golang-jwt/jwt/v4
> cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
> BenchmarkToken_SigningString
> BenchmarkToken_SigningString/BenchmarkToken_SigningString
> BenchmarkToken_SigningString/BenchmarkToken_SigningString-12         	  771712	      1400 ns/op	     600 B/op	      17 allocs/op
> PASS